src/GitServer/GitServer.ts

```diff
@@ -0,0 +1,298 @@
+    return fs.existsSync(Path.join(this.repoDir, repo))
+  }
+
+  /**
+   * returns the typeof service being process
+   */
+  getType(service: any): string {
+    switch (service) {
+      case 'upload-pack':
+        return 'fetch'
+      case 'receive-pack':
+        return 'push'
+      default:
+        return 'unknown'
+    }
+  }
+
+
+  // alternatively you can also pass authenticate a promise
+  authenticate(
+    type: string,
+    repo: string,
+    username: string,
+    password: string,
+    headers: http.IncomingHttpHeaders
+  ): boolean {
+    if(username === 'foo') {
+      return true
+    }
+    return false
+  }
+
+  /**
+   * Handle incoming HTTP requests with a connect-style middleware
+   */
+  handle(req: http.IncomingMessage, res: http.ServerResponse) {
+
+    const infoHandler = (req: http.IncomingMessage, res: http.ServerResponse): boolean => {
+
+      if (req.method !== 'GET') return false
+
+      const u = url.parse(req.url!)
+      const m = u.pathname!.match(/\/(.+)\/info\/refs$/)
+      if (!m) return false
+      if (/\.\./.test(m[1])) return false
+
+
+      const repo = m[1]
+      const params = parse(u.query!)
+      if (!params.service) {
+        res.statusCode = 400
+        res.end('service parameter required')
+        return true
+      }
+
+      const service = (<string>params.service).replace(/^git-/, '')
+      if (services.indexOf(service) < 0) {
+        res.statusCode = 405
+        res.end('service not available')
+        return true
+      }
+
+      const repoName = this.parseGitName(m[1])
+
+      // check if the repo is authenticated
+      const type = this.getType(service)
+      const headers = req.headers
+      const { username, password } = {username: 'foo', password: 's'}
+      let authenticated: boolean = false
+
+      if (username && password) {
+        authenticated = this.authenticate(type, repoName, username, password, headers)
+      }
+
+      if (!authenticated) {
+        res.setHeader("Content-Type", 'text/plain')
+        res.setHeader("WWW-Authenticate", 'Basic realm="authorization needed"')
+        res.writeHead(401)
+        res.end('Unauthorised to view this repo')
+        return true
+      }
+
+      this.infoResponse(repo, service, req, res)
+      return true
+    }
+    const requestPackHandler = (req: http.IncomingMessage, res: http.ServerResponse): boolean => {
+      if (req.method !== 'POST') return false
+      const m = req.url!.match(/\/(.+)\/git-(.+)/)
+      if (!m) return false
+      if (/\.\./.test(m[1])) return false
+
+      const repo = m[1]
+      const service = m[2]
+
+      if (services.indexOf(service) < 0) {
+        res.statusCode = 405
+        res.end('service not available')
+        return true
+      }
+
+      res.setHeader('content-type', 'application/x-git-' + service + '-result')
+      this.noCache(res)
+
+      const repoDir = Path.join(this.repoDir, repo)
+
+      const fileSystem = this.vaultStore.getVault(repo)?.efs
```

Comment:
What happens if the vault store doesn't have that repo? You used the `?` here to ignore the null. But where is it confirmed that such a repo indeed exists prior?

Reply:
I've added new code to the exists method, it should check 2 things;
1) does the vault exist? (this includes if it exists on file as well as if polykey has it in its stores, both need to be true); and
2) is the peer authorised to know it exists?

The last one is yet to be implemented and depends on getting the secure TLS connection right so we know who is connecting (client authentication)